File: docs/llama-router-preset-ui-execution/contracts/validate-contracts.py
#!/usr/bin/env python3
"""Phase 0 contract parse validation. Schema validation deferred to Phase 1A (jsonschema)."""
from __future__ import annotations

import json
import sys
from pathlib import Path

CONTRACTS_DIR = Path(__file__).resolve().parent
FIXTURES = sorted(CONTRACTS_DIR.glob("*.fixture.json"))
SCHEMA_FILES = sorted(CONTRACTS_DIR.glob("schema.*.json"))


def load_json(path: Path) -> object:
    with path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def main() -> int:
    failures: list[str] = []
    d12_path = CONTRACTS_DIR / "immutable-operation-input.fixture.json"

    for path in SCHEMA_FILES + FIXTURES:
        try:
            payload = load_json(path)
        except json.JSONDecodeError as exc:
            failures.append(f"{path.name}: JSON parse failed: {exc}")
            continue

        if not isinstance(payload, (dict, list)):
            failures.append(f"{path.name}: root must be object or array")

    if d12_path.exists():
        mtp = load_json(d12_path)
        assert isinstance(mtp, dict)
        if str(mtp.get("definitionId", "")).endswith("-mtp"):
            if not mtp.get("mmprojFiles"):
                failures.append(
                    "immutable-operation-input.fixture.json: MTP rows require mmprojFiles when vision is enabled"
                )
            preset = mtp.get("routerPreset") or {}
            if isinstance(preset, dict):
                if "image-min-tokens" not in preset:
                    failures.append(
                        "immutable-operation-input.fixture.json: MTP vision rows require routerPreset.image-min-tokens"
                    )
                if preset.get("spec-type") != "draft-mtp":
                    failures.append(
                        "immutable-operation-input.fixture.json: MTP rows require routerPreset.spec-type=draft-mtp"
                    )

    if failures:
        for item in failures:
            print(f"FAIL {item}", file=sys.stderr)
        return 1

    print(
        f"PASS parsed {len(FIXTURES)} fixtures and {len(SCHEMA_FILES)} schema files under {CONTRACTS_DIR.name}/"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: docs/llama-router-preset-ui-execution/contracts/validate-contracts.ts
import { readFileSync, readdirSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const contractsDir = dirname(fileURLToPath(import.meta.url));

interface CatalogGetResponse {
  schemaVersion: number;
  task: string;
  catalogVersion: string;
  models: unknown[];
}

interface ImmutableOperationInput {
  definitionId: string;
  definitionVersion: string;
  repository: string;
  resolvedRevision: string;
  modelFiles: string[];
  mmprojFiles: string[];
  routerModelId: string;
  runtimeProfileId: string;
  routerPreset: Record<string, string>;
}

const fixtureFiles = readdirSync(contractsDir)
  .filter((name) => name.endsWith('.fixture.json'))
  .sort();

const schemaFiles = readdirSync(contractsDir)
  .filter((name) => name.startsWith('schema.') && name.endsWith('.json'))
  .sort();

const failures: string[] = [];

for (const name of [...schemaFiles, ...fixtureFiles]) {
  const path = join(contractsDir, name);
  try {
    JSON.parse(readFileSync(path, 'utf8')) as unknown;
  } catch (error) {
    failures.push(`${name}: ${error instanceof Error ? error.message : String(error)}`);
  }
}

const catalog = JSON.parse(
  readFileSync(join(contractsDir, 'catalog-get-response.fixture.json'), 'utf8'),
) as CatalogGetResponse;
if (catalog.schemaVersion !== 1 || catalog.task !== 'llama') {
  failures.push('catalog-get-response.fixture.json: expected schemaVersion=1 task=llama');
}

const mtp = JSON.parse(
  readFileSync(join(contractsDir, 'immutable-operation-input.fixture.json'), 'utf8'),
) as ImmutableOperationInput;
if (mtp.definitionId.endsWith('-mtp')) {
  if (mtp.mmprojFiles.length === 0) {
    failures.push('immutable-operation-input.fixture.json: MTP vision rows require mmprojFiles');
  }
  if (!('image-min-tokens' in mtp.routerPreset)) {
    failures.push('immutable-operation-input.fixture.json: MTP vision rows require routerPreset.image-min-tokens');
  }
  if (mtp.routerPreset['spec-type'] !== 'draft-mtp') {
    failures.push('immutable-operation-input.fixture.json: MTP rows require routerPreset.spec-type=draft-mtp');
  }
}

if (failures.length > 0) {
  for (const failure of failures) {
    console.error(`FAIL ${failure}`);
  }
  process.exit(1);
}

console.log(
  `PASS parsed ${fixtureFiles.length} fixtures and ${schemaFiles.length} schema files under contracts/`,
);
